drivers/pn532/pn532.c

```diff
@@ -0,0 +1,696 @@
+}
+
+int pn532_write_reg(pn532_t *dev, unsigned addr, char val)
+{
+    int ret = -1;
+    char buff[PN532_BUFFER_LEN];
+
+    buff[BUFF_CMD_START     ] = PN532_CMD_WRITE_REG;
+    buff[BUFF_DATA_START    ] = (addr >> 8) & 0xff;
+    buff[BUFF_DATA_START + 1] = addr & 0xff;
+    buff[BUFF_DATA_START + 2] = val;
+
+    ret = send_rcv(dev, buff, 3, 0);
+
+    return ret;
+}
+
+int pn532_read_gpio(pn532_t *dev, uint32_t *gpio)
+{
+    int ret = -1;
+    char buff[PN532_BUFFER_LEN];
+
+    buff[BUFF_CMD_START     ] = PN532_CMD_READ_GPIO;
+
+    if (send_rcv(dev, buff, 0, 3) == 3) {
+        *gpio = (buff[1] << 16);    /* p3 */
+        *gpio += (buff[2] << 8);    /* p7 */
+        *gpio += (buff[3] << 0);    /* l0l1 */
+        ret = 0;
+    }
+
+    return ret;
+}
+
+int pn532_write_gpio(pn532_t *dev, unsigned p3, unsigned p7)
+{
+    int ret = -1;
+    char buff[PN532_BUFFER_LEN];
+
+    buff[BUFF_CMD_START     ] = PN532_CMD_WRITE_GPIO;
+    buff[BUFF_DATA_START    ] = (char)p3 | 0x80;
+    buff[BUFF_DATA_START + 1] = (char)p7 | 0x80;
+
+    ret = send_rcv(dev, buff, 2, 0);
+
+    return ret;
+}
+
+#define RF_CONFIG_RF_FIELD         (0x01)
+#define RF_CONFIG_TIMING           (0x02)
+#define RF_CONFIG_MAX_RTY_COM      (0x04)
+#define RF_CONFIG_MAX_RETRIES      (0x05)
+#define RF_CONFIG_ANALOG_A_106     (0x0A)
+#define RF_CONFIG_ANALOG_A_212_424 (0x0B)
+#define RF_CONFIG_ANALOG_B         (0x0C)
+#define RF_CONFIG_ANALOG_ISO1443_4 (0x0D)
+
+static int _rf_configure(pn532_t *dev, char *buff, unsigned cfg_item, char *config,
+                         unsigned cfg_len)
+{
+    buff[BUFF_CMD_START ] = PN532_CMD_RF_CONFIG;
+    buff[BUFF_DATA_START] = cfg_item;
+    for (int i = 1; i <= cfg_len; i++) {
+        buff[BUFF_DATA_START + i] = *config++;
+    }
+
+    return send_rcv(dev, buff, cfg_len + 1, 0);
+}
+
+static int _set_act_retries(pn532_t *dev, char *buff, unsigned max_retries)
+{
+    char rtrcfg[] = { 0xff, 0x01, max_retries & 0xff };
+
+    return _rf_configure(dev, buff, RF_CONFIG_MAX_RETRIES, rtrcfg, sizeof(rtrcfg));
+}
+
+int pn532_rf_configure(pn532_t *dev, unsigned cfg_item, char *config,
+                       unsigned cfg_len)
+{
+    char buff[PN532_BUFFER_LEN];
+
+    return _rf_configure(dev, buff, cfg_item, config, cfg_len);
+}
+
+int pn532_sam_configuration(pn532_t *dev, pn532_sam_conf_mode_t mode, unsigned timeout)
+{
+    int ret = -1;
+    char buff[PN532_BUFFER_LEN];
+
+    buff[BUFF_CMD_START     ] = PN532_CMD_SAM_CONFIG;
+    buff[BUFF_DATA_START    ] = (char)mode;
+    buff[BUFF_DATA_START + 1] = (char)(timeout / 50);
+    buff[BUFF_DATA_START + 2] = 0x01;
+
+    ret = send_rcv(dev, buff, 3, 0);
+
+    return ret;
+}
+
+static int _list_passive_targets(pn532_t *dev, char *buff, pn532_target_t target, unsigned max)
+{
+    buff[BUFF_CMD_START] = PN532_CMD_LIST_PASSIVE;
+    buff[BUFF_DATA_START] = (char) max;
+    buff[BUFF_DATA_START + 1] = (char)target;
+
+    /* requested len depends on expected target num and type */
```

Comment:
That's a good catch! 

20 works well for 1 card of type `PN532_BR_106_ISO_14443_A` (this is mifare and other iso14443a). 
Just checked the datasheet (page 116) and I'll modify slightly the code so it's clear what it currently supports.